### PR TITLE
Costing Unification: move from `factor_total_investment` to `TIC`

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
The two costing packages have different ways of accounting for indirect capital costs. The ZO Costing Package allows the indirect capital cost to be specified per unit model. The WT Costing Package has a global-level indirect capital cost.

This PR would resolve the conflict in favor of the ZO costing package. This requires modifying WaterTAP's existing detailed model's costing methods, and a few related changes to our detailed flowsheets.

## Changes proposed in this PR:
- Move `TIC` / `TPEC` factor calculation to the costing base class
- Add `TIC` factor to all unit models which do not already have a factor
- Specify a default o `2` for `TIC` when used in the WT Costing Package
- Make `factor_total_investment` have a default value of `1` in the WT Costing Package
- Various changes to detailed flowsheets to reproduce existing results.
- Set automatic scaling factors on unit model costing aggregates to resolve some resulting detailed flowsheet instability (in particular, NF and LSRRO).

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by avdudchenko — APPROVED** [review]
This looks good to me.

File: watertap/costing/unit_models/ion_exchange.py [hunk 1]
```diff
@@ -336,9 +336,11 @@ def cost_ion_exchange(blk):
             to_units=blk.costing_package.base_currency,
         )
     )
+    blk.costing_package.add_cost_factor(blk, "TIC")
     blk.capital_cost_constraint = pyo.Constraint(
         expr=blk.capital_cost
-        == pyo.units.convert(
+        == blk.cost_factor
```
Comment (kurbansitterley) [inline watertap/costing/unit_models/ion_exchange.py:337]: There is currently a ``total_installed_cost_factor`` with value of 1.65 used in capital costs for IX that I had included before there was a global ``TIC``. That should be removed entirely and use the global variable to not be double counting.

File: watertap/costing/unit_models/ion_exchange.py [hunk 2]
```diff
@@ -336,9 +336,11 @@ def cost_ion_exchange(blk):
             to_units=blk.costing_package.base_currency,
         )
     )
+    blk.costing_package.add_cost_factor(blk, "TIC")
```
Comment (kurbansitterley) [inline watertap/costing/unit_models/ion_exchange.py:334]:
```suggestion
    blk.costing_package.add_cost_factor(blk, None)
```
I suppose we could also just use the local IX ``total_installed_cost_factor`` doing something like this? I am fine with the system TIC also and would let you make the decision.

File: watertap/costing/unit_models/ion_exchange.py (hunk 1 above)
Comment (kurbansitterley) [reply watertap/costing/unit_models/ion_exchange.py:337]: Of course whatever you decide to do it will cause the IX test to fail.

File: watertap/unit_models/tests/test_anaerobic_digestor.py [hunk 3]
```diff
@@ -407,7 +407,7 @@ def test_costing(self, adm):
         assert_optimal_termination(results)
 
         # Check solutions
-        assert pytest.approx(1083290.8, rel=1e-5) == value(
+        assert pytest.approx(2.0 * 1083290.8, rel=1e-5) == value(
```
Comment (kurbansitterley) [inline watertap/unit_models/tests/test_anaerobic_digestor.py:413]: How does doubling the capital costs via this factor not affect LCOW? I don't see any LCOW tests changed.

**Review by kurbansitterley — CHANGES_REQUESTED** [review]
There is a change needed for the ion_exchange costing model to not double count indirect costs with the use of TIC. As far as I know, that is the only unit model that had a local factor to account or indirect costs.

File: watertap/costing/unit_models/ion_exchange.py (hunk 2 above)
Comment (bknueven) [reply watertap/costing/unit_models/ion_exchange.py:334]: Okay, thanks for catching this. I will use the system TIC and modify IX costing model appropriately.

File: watertap/unit_models/tests/test_anaerobic_digestor.py (hunk 3 above)
Comment (bknueven) [reply watertap/unit_models/tests/test_anaerobic_digestor.py:413]: The factor is now part of the unit model's `capital_cost` variable, as opposed to being applied at the global level.

Comment (kurbansitterley) [reply watertap/unit_models/tests/test_anaerobic_digestor.py:413]: Ok I see. 

Explanation for myself and anyone else: ``factor_capital_cost`` was set to 1 so mooted the influence of having ``TIC`` applied at the unit model level. So the value of ``aggregate_capital_cost``, which is the sum of unit model ``capital_cost``,  didn't change; ``aggregate_capital_cost``  is used in ``total_capital_cost``, which is used in ``LCOW``.

Comment (bknueven) [reply watertap/unit_models/tests/test_anaerobic_digestor.py:413]: For others, this change is being made to create uniformity around "where" the indirect capital cost is applied.

Right now the ZOCostingPackage does things at the unit-level, and the WaterTAPCosting package at the global-level. The PR modifies all the existing detailed models to put it at the unit-level.

**Comment by bknueven** [comment]
@kurbansitterley I believe I have addressed your concerns around the ion exchange model in 8264aae.

**Review by kurbansitterley — APPROVED** [review]
LGTM

**Comment by bknueven** [comment]
There are two costing PRs open, #1186 and #1196. Should this wait on those?

**Comment by adam-a-a** [comment]
> There are two costing PRs open, #1186 and #1196. Should this wait on those?

Not necessarily.

File: watertap/examples/flowsheets/lsrro/tests/test_lssro_paper_analysis.py [hunk 4]
```diff
@@ -86,18 +86,21 @@ def test_against_paper_analysis(csv_file, row_index):
...
+            **input_arguments,
+            has_NaCl_solubility_limit=True,
+            permeate_quality_limit=1000e-6,
+            has_calculated_concentration_polarization=True,
+            has_calculated_ro_pressure_drop=True,
+            A_value=5 / 3.6e11,
+            B_max=None,
+            number_of_RO_finite_elements=10
+        )
+    except ValueError:
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/lsrro/tests/test_lssro_paper_analysis.py:100]: Why was this needed? To sidestep the occasional fail related to this test or something else?

Comment (bknueven) [reply watertap/examples/flowsheets/lsrro/tests/test_lssro_paper_analysis.py:100]: If Ipopt terminates with a weird termination condition, then Pyomo will raise a `ValueError` trying to load the results back into the model.

This simply catches this mode of failure. Note that if the solve fails (or succeeds) and we didn't expect it to the test will still fail.

**Comment by bknueven** [comment]
@kurbansitterley could you have a look at 55e1754? I left a few things out on the previous commit...

**Comment by kurbansitterley** [comment]
> @kurbansitterley could you have a look at [55e1754](https://github.com/watertap-org/watertap/commit/55e1754e258e7f66b935d4a834f53c31e10f3c61)? I left a few things out on the previous commit...

Yes, see that. Most important change was addressed (not double counting indirect costs) and is now reflected in the tests, it appears. So my approval remains!

**Comment by bknueven** [comment]
This is currently held up by the failing notebook test `tutorials/nawi_spring_meeting2023.ipynb`.

@adam-a-a @avdudchenko any thoughts?

File: watertap/unit_models/tests/test_cstr.py [hunk 5]
```diff
@@ -456,16 +457,18 @@ def test_costing(self):
...
+        results = solver.solve(m, tee=True)
 
         assert_optimal_termination(results)
 
         # Check solutions
         assert pytest.approx(526.45 * 2, rel=1e-5) == value(
             m.fs.unit.costing.capital_cost
         )
-        assert pytest.approx(1.07948e-5, rel=1e-5) == value(m.fs.costing.LCOW)
+        assert pytest.approx(8.94365e-07, rel=1e-5) == value(m.fs.costing.LCOW)
```
Comment (bknueven) [inline watertap/unit_models/tests/test_cstr.py]: @MarcusHolly can you have a look at this? It seems like this flowsheet had degrees of freedom but no objective, and therefore the LCOW could converge to an arbitrary value.

I've fixed it here to what I got out of the optimization.

Comment (MarcusHolly) [reply watertap/unit_models/tests/test_cstr.py]: Sure, I can take a look at it soon, just not right now...

**Comment by bknueven** [comment]
@adam-a-a should this wait on #1213? 7d33365 might be of help with that, so we might want to merge this anyways.

**Comment by adam-a-a** [comment]
> This is currently held up by the failing notebook test `tutorials/nawi_spring_meeting2023.ipynb`.
> 
> @adam-a-a @avdudchenko any thoughts?

Is this still an issue or have your latest commits resolved this?

**Comment by adam-a-a** [comment]
> @adam-a-a should this wait on #1213? [7d33365](https://github.com/watertap-org/watertap/commit/7d33365e4285183cab6fa0c312f6942e49ab9efc) might be of help with that, so we might want to merge this anyways.

I suppose this doesn't need to wait on #1213 , assuming it won't break the BSM2 flowsheet.

**Comment by bknueven** [comment]
> > This is currently held up by the failing notebook test `tutorials/nawi_spring_meeting2023.ipynb`.
> > @adam-a-a @avdudchenko any thoughts?
> 
> Is this still an issue or have your latest commits resolved this?

No, that was fixed by 7d33365 (making LCOW an Expression)

**Comment by bknueven** [comment]
> > @adam-a-a should this wait on #1213? [7d33365](https://github.com/watertap-org/watertap/commit/7d33365e4285183cab6fa0c312f6942e49ab9efc) might be of help with that, so we might want to merge this anyways.
> 
> I suppose this doesn't need to wait on #1213 , assuming it won't break the BSM2 flowsheet.

Sounds like we should wait ... I mean, it shouldn't, but I am not extraordinarily confident either.

**Comment by bknueven** [comment]
Going to wait on #1213.
